### PR TITLE
Add basic byte size estimation for batches

### LINKED ISSUE #21655: batch-consuming DoFns should estimate byte size
Imported from Jira [BEAM-14408](https://issues.apache.org/jira/browse/BEAM-14408). Original Jira may contain additional context.
Reported by: bhulette.
Subtask of issue #21650

### PR DESCRIPTION
**Pull request by TheNeuralBit**
Fixes #21655

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/python/apache_beam/runners/worker/opcounters.py [hunk 1]
```diff
@@ -219,9 +219,14 @@ def update_from_batch(self, windowed_batch):
...
-        self.producer_batch_converter.get_length(windowed_batch.values))
-    # TODO(BEAM-14408): Update byte size estimate
+    batch_length = self.producer_batch_converter.get_length(
+        windowed_batch.values)
+    self.element_counter.update(batch_length)
+
+    mean_element_size = self.producer_batch_converter.estimate_byte_size(
+        windowed_batch.values) / batch_length
+    for _ in range(batch_length):
+      self.mean_byte_counter.update(mean_element_size)
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/runners/worker/opcounters.py:229]: This seems very naive, but I'm not sure of another way to keep this consistent with the element-wise behaviour, which is tracking the distribution of _element_ byte size estimates. Definitely open to suggestions here.

File: sdks/python/apache_beam/typehints/batch.py [hunk 2]
```diff
@@ -104,6 +110,12 @@ def __hash__(self) -> int:
 
 
 class ListBatchConverter(BatchConverter):
+  SAMPLE_FRACTION = 0.2
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/typehints/batch.py:113]: Similarly here, this is just an arbitrary sample rate. Should we try harder to keep this consistent with the reservoir sampling done in `OperationCounters`?

Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch.py:113]: Let's add an upper bound here.

File: sdks/python/apache_beam/runners/worker/opcounters.py (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/runners/worker/opcounters.py:229]: Look at adding an `update_n` to distribution counter

File: sdks/python/apache_beam/typehints/batch.py (hunk 2 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch.py:113]: Done.

**Comment by TheNeuralBit** [comment]
R: @robertwb PTAL, I've addressed the comments as we discussed offline

**Review by robertwb — APPROVED** [review]

**Comment by TheNeuralBit** [comment]
Run Portable_Python PreCommit

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

**Comment by TheNeuralBit** [comment]
Run PythonDocker PreCommit

**Comment by TheNeuralBit** [comment]
Run Python_PVR_Flink PreCommit

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

**Comment by TheNeuralBit** [comment]
Run Python_PVR_Flink PreCommit

**Comment by TheNeuralBit** [comment]
Run Python_PVR_Flink PreCommit

**Pull request merged**